kong/plugins/moesif/moesif_gov.lua

```diff
@@ -0,0 +1,227 @@
+        fetch_governance_rules[hash_key] = true
+        if conf.debug then
+            ngx_log(ngx_log_ERR, "[moesif] failed to keepalive to " .. parsed_url.host .. ":" .. tostring(parsed_url.port) .. ": ", err)
+        end
+      return
+     else
+      if conf.debug then
+        ngx_log(ngx.DEBUG,"[moesif] success keep-alive", ok)
+      end
+    end
+
+    local governance_rules = {}
+    local response_body = cjson.decode(governance_rules_response:match("(%[.*])"))
+    for k, rule in pairs(response_body) do
+        governance_rules[rule["_id"]] = rule
+    end
+
+    if type(governance_rules) == "table" and next(governance_rules) ~= nil then
+        governance_rules_hashes[hash_key] = governance_rules
+    end
+
+    local rules_etag = string.match(governance_rules_response, "Rules%s*-%s*ETag%s*:%s*(.-)\n")
+    governance_rules_etags[hash_key] = rules_etag
+
+    -- Set fetch_governance_rule to true, to be able to fetch governance rule next time
+    fetch_governance_rules[hash_key] = true
+  end
+
+-- Check if need to block request based on the governance rule of the entity associated with the request
+-- @param hash_key    Hash key of the config application Id
+-- @param `conf`      Configuration table, holds http endpoint details
+-- @param `rule_name` User or Company 
+-- @param `entity_id` User or Company Id associated with the reqeust
+function block_request_based_on_entity_governance_rule(hash_key, conf, rule_name, entity_id)
+    if governance_rules_hashes[hash_key] ~= nil and type(governance_rules_hashes[hash_key]) == "table" and next(governance_rules_hashes[hash_key]) ~= nil 
+        and conf[rule_name] ~= nil and conf[rule_name][entity_id] ~= nil then 
+
+        local governance_rules = governance_rules_hashes[hash_key]
+        local entity_rule = conf[rule_name][entity_id]
+        if entity_rule[0] ~= nil then
+            entity_rule = entity_rule[0]
+        else
+            entity_rule = entity_rule[1]
+        end
+        
+        local rule_id = entity_rule["rules"]
+        if rule_id ~= nil and governance_rules[rule_id] ~= nil then 
+            local governance_rule = governance_rules[rule_id]
+            local is_block = governance_rule["block"]
+
+            if is_block and governance_rule["response"] ~= nil and governance_rule["response"]["status"] ~= nil 
+                and governance_rule["response"]["headers"] ~=nil and governance_rule["response"]["body"] ~= nil then
+                local gr_status = governance_rule["response"]["status"]
+                local gr_headers = governance_rule["response"]["headers"]
+                local gr_body = governance_rule["response"]["body"]
+                local rule_values = entity_rule["values"]
+
+                local updated_gr_headers = {}
+                for k,v in pairs(gr_headers) do updated_gr_headers[k]=v end
+
+                if rule_values ~= nil and type(updated_gr_headers) == "table" and next(updated_gr_headers) ~= nil then
```

Comment:
Yes, it will always be table but it is best practice to check if the variable is of type table before checking if table is empty or not. We could remove it if you want.

Reply:
its fine